Sources/SwiftSyntaxBuilder/SyntaxNodeWithBody.swift

```diff
@@ -124,7 +126,17 @@ extension WithOptionalCodeBlockSyntax where Self: DeclSyntaxProtocol {
     _ header: SyntaxNodeString,
     @CodeBlockItemListBuilder bodyBuilder: () throws -> CodeBlockItemListSyntax
   ) throws {
-    let decl = DeclSyntax("\(header) {}")
+    // If the type provides a custom `SyntaxParseable` implementation, use that. Otherwise construct it as a
+    // `DeclSyntax`.
+    let decl: DeclSyntax
+    var stringInterpolation = SyntaxStringInterpolation(literalCapacity: 1, interpolationCount: 1)
+    stringInterpolation.appendInterpolation(header)
+    stringInterpolation.appendLiteral(" {}")
+    if let parsableType = Self.self as? SyntaxParseable.Type {
+      decl = parsableType.init(stringInterpolation: stringInterpolation).cast(DeclSyntax.self)
```

Comment:
I think upcasting here and down casting again below is a waste. Instead of branching in `where Self: DeclSyntaxProtocol` extension, how about making `where Self: SyntaxParseable` separately?
```swift
extension WithOptionalCodeBlockSyntax where Self: SyntaxParseable & DeclSyntaxProtocol { ... 
```
(`& DeclSyntaxProtocol` is for resolving ambiguity)

Reply:
Discussed offline and kept the current implementation + added a comment why we need to manually construct the string interpolation.